lib/ansible/galaxy/collection/gpg.py

```diff
@@ -0,0 +1,269 @@
+    from typing import TYPE_CHECKING
+except ImportError:
+    TYPE_CHECKING = False
+
+if TYPE_CHECKING:
+    from ansible.utils.display import Display
+    from typing import Tuple, Iterator
+
+
+IS_PY310_PLUS = sys.version_info[:2] >= (3, 10)
+
+frozen_dataclass = partial(dataclass, frozen=True, **({'slots': True} if IS_PY310_PLUS else {}))
+
+
+def run_gpg_verify(
+    manifest_file,  # type: str
+    signature,  # type: str
+    keyring,  # type: str
+    display,  # type: Display
+):  # type: (...) -> Tuple[str, int]
+    status_fd_read, status_fd_write = os.pipe()
+
+    # running the gpg command will create the keyring if it does not exist
+    remove_keybox = not os.path.exists(keyring)
+
+    cmd = [
+        'gpg',
+        f'--status-fd={status_fd_write}',
+        '--verify',
+        '--batch',
+        '--no-tty',
+        '--no-default-keyring',
+        f'--keyring={keyring}',
+        '-',
+        manifest_file,
+    ]
+    cmd_str = ' '.join(cmd)
+    display.vvvv(f"Running command '{cmd}'")
+
+    try:
+        p = subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            pass_fds=(status_fd_write,),
+            encoding='utf8',
+        )
+    except (FileNotFoundError, subprocess.SubprocessError) as err:
+        raise AnsibleError(
+            f"Failed during GnuPG verification with command '{cmd_str}': {err}"
+        ) from err
+    else:
+        stdout, stderr = p.communicate(input=signature)
+    finally:
+        os.close(status_fd_write)
+
+    if remove_keybox:
+        try:
+            os.remove(keyring)
+        except OSError:
+            pass
+
+    with os.fdopen(status_fd_read) as f:
+        stdout = f.read()
+        display.vvvv(f"{stdout} (exit code {p.returncode})")
+        return stdout, p.returncode
+
+
+def parse_gpg_errors(status_out: str):  # -> Iterator[GpgBaseError]
+    for line in status_out.splitlines():
+        if not line:
+            continue
+        try:
+            _dummy, status, remainder = line.split(maxsplit=2)
+        except ValueError:
+            _dummy, status = line.split(maxsplit=1)
+            remainder = None
+
+        try:
+            cls = GPG_ERROR_MAP[status]
+        except KeyError:
+            continue
+
+        fields = [status]
+        if remainder:
+            fields.extend(
+                remainder.split(
+                    None,
+                    len(dc_fields(cls)) - 2
+                )
+            )
+
+        yield cls(*fields)
+
+
+@frozen_dataclass
+class GpgBaseError(Exception):
+    status: str
+
+    @classmethod
+    def get_gpg_error_description(cls) -> str:
+        """Return the current class description."""
+        return ' '.join(cls.__doc__.split())
+
+    def __post_init__(self):
+        for field in dc_fields(self):
+            super(GpgBaseError, self).__setattr__(field.name, field.type(getattr(self, field.name)))
```

Comment:
Is this `super(GpgBaseError, self)` needed instead of `super()` because it's called multiple times, and each time it hits a different superclass per MRO?
I wonder if it could be like `set_super_attr = super().__setattr__` before the loop that would be used as
```suggestion
            set_super_attr(field.name, field.type(getattr(self, field.name)))
```
here. Would that work?

Also, it's a bit complicated to guess what this does. Mind adding a code comment with some motivation behind the trick?

Reply:
I think this method is actually unnecessary so I've tentatively removed it.... Not sure what the motivation behind the trick was as I didn't write it (was part of a POC sivel put together).

It actually was failing on the first loop. Somehow the obj is not an instance of the base class. I tried to make an isolated reproducer and couldn't (the ansible-galaxy-collection tests triggered the bug though). And it seems unrelated to the loop - setting `set_super_attr = super().__setattr__` before the loop and using that made no difference.